openedx/openedx-platform
# Cleaning up base.in: Remove unused dependencies, separate optional ones

## Description

The platform's dependency list is very long and I'm interested in working to simplify it. This PR is focused on [`base.in`](https://github.com/openedx/edx-platform/blob/master/requirements/edx/base.in) and it does three things:

1. **Remove unused dependencies**:
    * `contextlib2` is not required anymore now that we're on Python 3
    * `docopt` was used only for parsing/documenting args for a command-line script (`static_content.py`) that accepted a single argument. I've refactored it to not use `docopt`.
    * `python-Levenshtein` did not appear to be used at all.
2. **Remove indirect dependencies**. The instructions at the top of `base.in` say, "Do not add indirect dependencies unless a version constraint is needed to avoid versions which have known problems." Yet several indirect dependencies were specified, so I removed them:
    * `django-appconf` (dependency of `django-statici18n`)
    * `django-fernet-fields` (dependency of `edx-enterprise`)
    * `nltk` (dependency of `chem`)
    * `super-csv` (dependency of `edx-bulk-grades`)
3. **Separate out "non-kernel" dependencies**. As discussed in [A minimal Open edX distribution](https://discuss.openedx.org/t/a-minimal-open-edx-distribution/9478), I want to get the process started of separating "kernel" dependencies (required for the Core Product / Open edX platform to run) and other dependencies like XBlocks and plugins that are not considered "Core" or that are only used on edx.org. In order to work toward that goal incrementally, for now the only thing I have done is moved them to a separate region of `base.in` and added comments saying that they look like "non-core" dependencies. In a future PR, I want to move these to a new `default.in` file, which should still be used when installing the platform but which should not be installed for tests or for anyone who want a "minimal" build of Open edX.


The "non-kernel" dependencies I have tentatively identified for now are:

Third party integrations:
```
algoliasearch                       # Algolia’s API client for indexed searching
django-ses                          # Django email backend for Amazon’s Simple Email Service
edx-braze-client                    # a customer engagement platform used for edx.org
mailsnake                           # MailChimp API; used for two management commands in the "mailing" djangoapp
newrelic                            # New Relic agent for performance monitoring
optimizely-sdk                      # Optimizely provides A/B testing and other features, used by edx.org
```

Plugins used on edx.org:
```
learner-pathway-progress            # A plugin for lms to track learners progress in pathays
outcome-surveys                     # plugin "to send follow up events for learners who have passed a course"
skill-tagging                       # Library to support tagging of skills on edx.org in unit/video xblocks via forms
```

XBlocks:
```
crowdsourcehinter-xblock
done-xblock                         # a very simple XBlock that allows learners to mark an activity as Done
recommender-xblock                  # https://github.com/edx/RecommenderXBlock
staff-graded-xblock                 # https://github.com/openedx/staff_graded-xblock Allows off-site bulk scoring.
edx-sga                             # The more well known "staff graded assignment" XBlock, from MIT.
ora2>=4.5.0                         # Open Response Assessment XBlock
xblock-poll                         # Xblock for polling users
xblock-drag-and-drop-v2             # Drag and Drop XBlock
xblock-google-drive                 # XBlock for google docs and calendar
```

As well:
* `pygments` was only used to make paver output colorized, and paver will work just fine if it's not installed (and I think we want to get rid of paver?).
* `py2neo` is used by CourseGraph tasks/management commands, and CourseGraph is certainly non-core.

## Supporting information

* Discussion: [A minimal Open edX distribution?](https://discuss.openedx.org/t/a-minimal-open-edx-distribution/9478)
* [OEP 57 Component Tiers / defining the Kernel](https://docs.openedx.org/projects/openedx-proposals/en/latest/processes/oep-0057-proc-core-product.html#component-tiers)

## Testing instructions

For now, make sure tests pass and major functionality seems unaffected. The CI build should accomplish this.

## Deadline

None

## Other information

This does not have any immediate impact for Open edX developers nor operators. If and when we move the flagged dependencies to `default.in`, it will again have no immediate impact, but developers and operators will have a new choice of starting with the "minimal" build (kernel only) or "default" build as they prefer.
